eZ/Publish/Core/Persistence/Cache/ContentHandler.php

```diff
@@ -283,19 +301,38 @@ public function publish($contentId, $versionNo, MetadataUpdateStruct $struct)
     {
         $this->logger->logCall(__METHOD__, array('content' => $contentId, 'version' => $versionNo, 'struct' => $struct));
         $content = $this->persistenceHandler->contentHandler()->publish($contentId, $versionNo, $struct);
+        $this->cache->invalidateTags(['content-' . $contentId]);
 
-        $this->cache->clear('content', $contentId);
-        $this->cache->clear('content', 'info', 'remoteId');
-        $this->cache->clear('location', 'subtree');
+        return $content;
+    }
 
-        // warm up cache
-        $contentInfo = $content->versionInfo->contentInfo;
-        $this->cache
-            ->getItem('content', $contentInfo->id, $content->versionInfo->versionNo, self::ALL_TRANSLATIONS_KEY)
-            ->set($content)
-            ->save();
-        $this->cache->getItem('content', 'info', $contentInfo->id)->set($contentInfo)->save();
+    /**
+     * Return relevant content and location tags so cache can be purged reliably.
+     *
+     * @param ContentInfo $contentInfo
+     * @param bool $withFields Set to true if item contains fields which should be expired on relation or type updates.
+     * @param array $tags Optional, can be used to specify other tags.
+     *
+     * @return array
+     */
+    private function getCacheTags(ContentInfo $contentInfo, $withFields = false, array $tags = [])
```

Comment:
Wouldn't be better to create injectable kind of _ValueObject Cache Tagger_ helper (so it could be also used in extended/overwritten handlers without logic duplication) while still benefit from kernel cache invalidation in various places? (or is it planned responsibility of upcoming cache layer?)